Fix multistore reserved products state with shared catalog enabled

ComonSoft opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/8/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/8/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch           | develop / 8.1.x / 8.0.x / 1.7.8.x / 1.7.6.x
| Description      | This commit is to fix a bug in multistore mode with shared catalog, that decrement stock instead of place product as reserved until status set them to "Set the order as shipped". We fixed it by testing if multishop is active and shared stock is active to current shop group. In this case the sum must be made with the condition of shop_goup_id instead of shop_id for reserved product. And for physical quantity the shop_id must be 0 in this case. We tested from PS 1.7.6.4 to 8.1.x.
| Type             | bug fix 
| Category         | BO 
| BC breaks        |  no
| Deprecations     | no
| How to test      | Activate multistore, create a shop group with shared catalog, Add a shop in group. Place orders on each shop, with same and different products. Go to BO in Stocks, search for ordered products, you'll see them as reserved. Change orders status as In transit or delivred in both shops. Go to BO in Stocks search for ordered products, you'll see them without reserved and stock decremented.
| Fixed ticket     | Fixes #13210
| Related PRs       | no
| Sponsor company   | COM'ONSOFT @ComonSoft

DamFromCsoft committed:
Update StockManager.php

This commit fixes the issue  #13210 about the stock management in multistore mode with shared catalog.

kpodemski commented on src/Adapter/StockManager.php:
```diff
@@ -102,6 +103,11 @@ public function isAsmGloballyActivated()
         $this->updateReservedProductQuantity($shopId, $errorState, $cancellationState, $idProduct, $idOrder);
+        $whereShopIdCond = $shopId;
+
+        if(Shop::isFeatureActive() && Shop::getContextShopGroup()->share_stock){
```
You can use `ShopAdapter` class here and do `if ($shop_group->share_stock) {` take a look at the code in above methods. Also, no need to check if multi-store is enabled there, I think that will be handled anyway.

kpodemski commented on src/Adapter/StockManager.php:
```diff
@@ -170,7 +183,7 @@ private function updateReservedProductQuantity($shopId, $errorState, $cancellati
                 GROUP BY od.product_id, od.product_attribute_id
             )
-            WHERE sa.id_shop = :shop_id
+            WHERE sa.id_shop = '. $whereShopIdCond .'
```
```suggestion
            WHERE sa.id_shop = ' . $whereShopIdCond . '
```

kpodemski commented on src/Adapter/StockManager.php:
```diff
@@ -160,7 +173,7 @@ private function updateReservedProductQuantity($shopId, $errorState, $cancellati
                 INNER JOIN {table_prefix}order_detail od ON od.id_order = o.id_order
                 INNER JOIN {table_prefix}order_state os ON os.id_order_state = o.current_state
-                WHERE o.id_shop = :shop_id AND
+                WHERE '.$whereCondition.'
```
```suggestion
                WHERE ' . $whereCondition . '
```

kpodemski reviewed (commented):
cs fixer issues

ComonSoft commented:
Hi,

 

Do we need to make a new pull request or change files in this request?

 

Best Regards

Jean

 

De : Krystian Podemski ***@***.***> 
Envoyé : vendredi 26 mai 2023 11:00
À : PrestaShop/PrestaShop ***@***.***>
Cc : ComonSoft ***@***.***>; Author ***@***.***>
Objet : Re: [PrestaShop/PrestaShop] Fix multistore reserved products state with shared catalog enabled (PR #32721)

 

@kpodemski commented on this pull request.

  _____  

In src/Adapter/StockManager.php <https://github.com/PrestaShop/PrestaShop/pull/32721#discussion_r1206459113> :

> @@ -102,6 +103,11 @@ public function isAsmGloballyActivated()
     public function updatePhysicalProductQuantity($shopId, $errorState, $cancellationState, $idProduct = null, $idOrder = null)
     {
         $this->updateReservedProductQuantity($shopId, $errorState, $cancellationState, $idProduct, $idOrder);
+        $whereShopIdCond = $shopId;
+
+        if(Shop::isFeatureActive() && Shop::getContextShopGroup()->share_stock){

You can use ShopAdapter class here and do if ($shop_group->share_stock) { take a look at the code in above methods. Also, no need to check if multi-store is enabled there, I think that will be handled anyway.

—
Reply to this email directly, view it on GitHub <https://github.com/PrestaShop/PrestaShop/pull/32721#pullrequestreview-1445684710> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADHBX25NW46CHHCXG2WS26TXIBWJPANCNFSM6AAAAAAYP544QE> .
You are receiving this because you authored the thread.Message ID: ***@***.*** ***@***.***> >

kpodemski commented:
@ComonSoft you can continue editing the files on your branch :)

ComonSoft commented:
Hi Krystian,

 

I understand, but we implemented like this to be retro compatible with version from 1.7.6.4 to 8.1.x. Unfortunately your solution doesn’t works for PS version  less than 1.7.7.x ☹

 

 

De : Krystian Podemski ***@***.***> 
Envoyé : vendredi 26 mai 2023 11:00
À : PrestaShop/PrestaShop ***@***.***>
Cc : ComonSoft ***@***.***>; Author ***@***.***>
Objet : Re: [PrestaShop/PrestaShop] Fix multistore reserved products state with shared catalog enabled (PR #32721)

 

@kpodemski commented on this pull request.

  _____  

In src/Adapter/StockManager.php <https://github.com/PrestaShop/PrestaShop/pull/32721#discussion_r1206459113> :

> @@ -102,6 +103,11 @@ public function isAsmGloballyActivated()
     public function updatePhysicalProductQuantity($shopId, $errorState, $cancellationState, $idProduct = null, $idOrder = null)
     {
         $this->updateReservedProductQuantity($shopId, $errorState, $cancellationState, $idProduct, $idOrder);
+        $whereShopIdCond = $shopId;
+
+        if(Shop::isFeatureActive() && Shop::getContextShopGroup()->share_stock){

You can use ShopAdapter class here and do if ($shop_group->share_stock) { take a look at the code in above methods. Also, no need to check if multi-store is enabled there, I think that will be handled anyway.

—
Reply to this email directly, view it on GitHub <https://github.com/PrestaShop/PrestaShop/pull/32721#pullrequestreview-1445684710> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADHBX25NW46CHHCXG2WS26TXIBWJPANCNFSM6AAAAAAYP544QE> .
You are receiving this because you authored the thread.Message ID: ***@***.*** ***@***.***> >

kpodemski commented:
There's no need to think about backward compatibility if you improve a new version of the software :) This is not how it works. You always use the best means available to fix the problem 👍🏻

ComonSoft commented:
Ok, but there are hundreds of shops running with old versions, and this fixes a big bug open since 4 years, you mean they don’t need it?

kpodemski commented:
@ComonSoft but how would they implement the bug fix? the chance for them to have this bug fix in their stores is:
- upgrade to the new version
- implement the code changes manually

There won't be 1.7.6.something, 1.7.7.something, etc. We work on PrestaShop 8.1.x and 9.0.x versions.

DamFromCsoft committed:
Apply suggestions from code review

Co-authored-by: Krystian Podemski <kpodemski@users.noreply.github.com>

DamFromCsoft committed:
Apply suggestions from code review

Co-authored-by: Krystian Podemski <kpodemski@users.noreply.github.com>

DamFromCsoft committed:
Fix multistore reserved products state with shared catalog enabled #32721 

Creation of a method to get the shop_group according to the current shop using the shopAdapter class, and then test the share stock attribute.

DamFromCsoft committed:
Fix multistore reserved products state with shared catalog enabled #32721

ocio87 commented:
Hello, why did you guys stop this upgrade? It's excellent. Corrects the problem that has been occurring when there is a multi-store sharing inventory.

Tests are performed as follows:

Available stock of the product = 61.
Buy Store 1 = 1 unit
Buy Store 2 = 2 Units.

It is evident that it correctly reserves the units:

![image](https://github.com/PrestaShop/PrestaShop/assets/12537826/30a1e3af-f2d4-4678-9b0e-ae2832083588)

The first order for a unit from store 1 is canceled and inventory is correctly reduced:

![image](https://github.com/PrestaShop/PrestaShop/assets/12537826/b949fb1f-8947-4437-86b0-0669e56a436a)

![image](https://github.com/PrestaShop/PrestaShop/assets/12537826/aed26a59-a9bb-48bd-b344-892696049ee5)

The order in store 2 is changed to delivered and shipped and the inventory is correctly reduced. The unit of the canceled order is restored and the 2 units of store 2 are correctly reduced:

![image](https://github.com/PrestaShop/PrestaShop/assets/12537826/cc8e45bb-8d6d-454a-a2fa-14fe8d54b169)

ocio87 commented:
Hello, why did you guys stop this upgrade? It's excellent. Corrects the problem that has been occurring when there is a multi-store sharing inventory.

Prestashop version : 8.1.1
Php version: 8.1


Tests are performed as follows:

Available stock of the product = 61.
Buy Store 1 = 1 unit
Buy Store 2 = 2 Units.

It is evident that it correctly reserves the units:

![image](https://github.com/PrestaShop/PrestaShop/assets/12537826/42efd384-c449-4508-b152-fe0c4ff736ec)

The first order for a unit from store 1 is canceled and inventory is correctly reduced:

![image](https://github.com/PrestaShop/PrestaShop/assets/12537826/31460f3e-07df-4609-bc66-ad72cb161c63)

![image](https://github.com/PrestaShop/PrestaShop/assets/12537826/89a02707-1118-4d22-8728-6b8ff9aa52ef)


The order in store 2 is changed to delivered and shipped and the inventory is correctly reduced. The unit of the canceled order is restored and the 2 units of store 2 are correctly reduced:


![image](https://github.com/PrestaShop/PrestaShop/assets/12537826/47bc0fcc-9045-4664-a6f6-2a8a32fba86e)

![image](https://github.com/PrestaShop/PrestaShop/assets/12537826/4f5a5a7a-175b-47f6-b860-668f578bafde)

If I leave prestashop with the file unchanged, the inventory is damaged.

tom-combet commented:
> Hi, I don't understand the code issue, the method we added return ShopGroup object in any case, why the analyzer said the type should be PrestaShop\PrestaShop\Adapter\ShopGroup?
> 
> Method PrestaShop\PrestaShop\Adapter\StockManager::getShopGroup() should return PrestaShop\PrestaShop\Adapter\ShopGroup but returns ShopGroup.

Hi, looks like you have to either:
- add a slash to tell that ShopGroup comes from "Global Namespace" rather than a custom one
- import the class with a use statement at beginning of the file like this:
`use ShopGroup;`

I personally prefer the first way, since global namespace will ultimately disappear from PrestaShop, but both are valid.
